src/main/java/org/apache/cassandra/distributed/api/LogAction.java

```diff
@@ -0,0 +1,382 @@
+        Pattern exceptionPattern = Pattern.compile("[Ee]xception|AssertionError");
+        Function<String, String> extractLogLevel = line -> {
+            Matcher matcher = logLevelPattern.matcher(line);
+            if (!matcher.find())
+                return null;
+            return matcher.group(1);
+        };
+        List<String> matches = new ArrayList<>();
+        try (LineIterator it = match(startPosition, ignore -> true))
+        {
+            StringBuilder lineBuffer = new StringBuilder();
+            while (it.hasNext())
+            {
+                String line = it.next();
+                String logLevelOrNull = extractLogLevel.apply(line);
+                if (logLevelOrNull == null)
+                {
+                    // found a log which isn't the start of a logger line; assume its an exception
+                    if (lineBuffer.length() == 0)
+                    {
+                        // no previous start of line found, so skip this
+                        continue;
+                    }
+                    lineBuffer.append('\n').append(line);
+                    continue;
+                }
+                // line is a start of a log, reset state
+                if (lineBuffer.length() != 0)
+                {
+                    // buffer has content, add and move on
+                    matches.add(lineBuffer.toString());
+                    lineBuffer.setLength(0);
+                }
+                switch (logLevelOrNull)
+                {
+                    case "ERROR":
+                        lineBuffer.append(line);
+                        break;
+                    case "WARN":
+                        if (exceptionPattern.matcher(line).find())
+                            lineBuffer.append(line);
+                        break;
+                    default:
+                        // ignore
+                }
+            }
+            if (lineBuffer.length() != 0)
+            {
+                matches.add(lineBuffer.toString());
+            }
+            return new BasicLogResult<>(it.mark(), matches);
+        }
+    }
+
+    default LogResult<List<String>> grepForErrors()
+    {
+        return grepForErrors(Internal.DEFAULT_START_POSITION);
+    }
+
+    class BasicLogResult<T> implements LogResult<T>
+    {
+        private final long mark;
+        private final T result;
+
+        public BasicLogResult(long mark, T result) {
+            this.mark = mark;
+            this.result = Objects.requireNonNull(result);
+        }
+
+        @Override
+        public long getMark() {
+            return mark;
+        }
+
+        @Override
+        public T getResult() {
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "LogResult{" +
+                    "mark=" + mark +
+                    ", result=" + result +
+                    '}';
+        }
+    }
+    
+    class Internal
```

Comment:
We don't seem to need an internal class here, or? It's just a several additional helper methods.

Reply:
helper methods I don't want public.  In java 8 I can't have private fields or methods, so using a class with private fields/methods lets me hind things from the implementation.